### PR TITLE
INSP: refactor checking and filling of match expressions

### LINKED ISSUE #4349: Pattern matching LHS checker bad bad
## Environment

* **IntelliJ Rust plugin version:** 0.2.105.2728-192-nightly
* **Rust toolchain version:** 1.37.0 (eae3437df 2019-08-13) x86_64-pc-windows-msvc
* **IDE name and version:** CL-192.6262.62
* **Operating system:** Windows 10 10.0

## Problem description

image

This function should compile but intellij rust is too silly to realize that.

Code for copy&paste:

```rust
fn test() {
    let a = &Some(1);
    let b = &Some(1);
    match (a, b) {
        (None(..), None(..)) => (),
        (Some(..), Some(..)) => (),
        (Some(..), _) | (_, Some(..)) => (),
    }
}
```

## Fun fact

Removing `&` which makes `a` and `b` not references suppresses the error hint.

### LINKED ISSUE #5215: Rethink `Match check` inspection and `Fill match arms` intention
Currently, `Match check` inspection only work on non-empty `match` bodies, while  `Fill match arms` intention only work on empty ones. This is a bit strange. Probably they can be merged into one thing.

### PR DESCRIPTION
**Pull request by ortem**
Closes #5215
Fixes #4349

* Split `Match Check` into two inspections:
  - `RsNonExhaustiveMatchInspection` (E0004 error)
  - `RsUselessMatchArmInspection` (unreachable_patterns warning)
* Implement all `FillMatchArmsIntention` logic in `RsNonExhaustiveMatchInspection` and get rid of it

### CODE REVIEW

**Comment by Kobzol** [comment]
(https://github.com/intellij-rust/intellij-rust/issues/5215 is going to be the "tallest" issue in this repo soon, even though it does not have any comments :D )

File: src/main/resources/inspectionDescriptions/RsUselessMatchArm.html [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<html>
+<body>
+Checks that <code>match</code> expression doesn't have useless (unreachable) arms.
```
Comment (Undin) [inline src/main/resources/inspectionDescriptions/RsUselessMatchArm.html:3]: Note, it also detects unreachable patterns if arm uses or_patterns. So let's reword it to `Checks that <code>match</code> expression doesn't have useless (unreachable) patterns.`

**Review by Undin — COMMENTED** [review]
In general looks ok
Looks like compiler doesn't use `useless` word in warnings, only in code. It uses `unreachable` term. Also, the corresponding lint attribute (`unreachable_patterns`) also uses `unreachable`. Maybe it's worth to use `unreachable` as well? Note, we already use it in error annotations but inspection name and quick fixes use `useless` term.

A user even can see bot terms at the same time and it looks inconsistent
image

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                val separatorRange = (armPat.getPrevNonCommentSibling() as? LeafPsiElement)
+                    ?.takeIf { it.elementType == OR }
+                    ?.rangeWithPrevSpace
+                    ?: TextRange.EMPTY_RANGE
+
+                val range = armPat.rangeWithPrevSpace.union(separatorRange)
+                SubstituteTextFix.delete("Remove useless pattern", match.containingFile, range)
+            }
+
+            holder.registerProblem(armPat, "Unreachable pattern", ProblemHighlightType.WARNING, fix)
```
[the remark this answers is not in the document]
Comment (ortem) [reply src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt:72]: @Undin Do you think we should use "unused" highlighting (gray code) instead of "warning" highlighting (yellow background code)? I've chosen `LIKE_UNUSED_SYMBOL` in `RsLint.UnreachablePattern` at first, but I'm not sure because unreachable match arms are more likely to indicate a programmer's mistake than just unused variables

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/RsNonExhaustiveMatchInspection.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import org.rust.lang.core.psi.RsVisitor
+import org.rust.lang.core.psi.ext.arms
+import org.rust.lang.core.types.infer.containsTyOfClass
+import org.rust.lang.core.types.ty.TyUnknown
+import org.rust.lang.core.types.type
+import org.rust.lang.utils.RsDiagnostic
+import org.rust.lang.utils.addToHolder
+
+class RsNonExhaustiveMatchInspection : RsLocalInspectionTool() {
+    override fun getDisplayName() = "Non-exhaustive match"
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/checkMatch/RsNonExhaustiveMatchInspection.kt:20]: nit: explicit type?

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/RsNonExhaustiveMatchInspection.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            if (exprType.containsTyOfClass(TyUnknown::class.java)) return
+            try {
+                checkExhaustive(matchExpr, holder)
+            } catch (todo: NotImplementedError) {
+            } catch (e: CheckMatchException) {
+            }
+        }
+    }
+
+    private fun checkExhaustive(match: RsMatchExpr, holder: RsProblemsHolder) {
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/checkMatch/RsNonExhaustiveMatchInspection.kt:34]:
```suggestion
    private fun checkExhaustiveness(match: RsMatchExpr, holder: RsProblemsHolder) {
```

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import org.rust.lang.core.psi.RsMatchArm
+import org.rust.lang.core.psi.RsMatchExpr
+import org.rust.lang.core.psi.RsVisitor
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.types.infer.containsTyOfClass
+import org.rust.lang.core.types.ty.TyUnknown
+import org.rust.lang.core.types.type
+
+class RsUselessMatchArmInspection : RsLintInspection() {
+    override fun getDisplayName() = "Unreachable match arm"
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt:25]: nit: explicit type?

Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt:25]:
```suggestion
    override fun getDisplayName() = "Unreachable match patterns"
```

or even
```suggestion
    override fun getDisplayName() = "Unreachable patterns"
```

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+import org.rust.lang.core.psi.RsElementTypes.OR
+import org.rust.lang.core.psi.RsMatchArm
+import org.rust.lang.core.psi.RsMatchExpr
+import org.rust.lang.core.psi.RsVisitor
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.types.infer.containsTyOfClass
+import org.rust.lang.core.types.ty.TyUnknown
+import org.rust.lang.core.types.type
+
+class RsUselessMatchArmInspection : RsLintInspection() {
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt:24]:
```suggestion
class RsUselessMatchPatternInspection : RsLintInspection() {
```

File: src/main/kotlin/org/rust/ide/inspections/lints/RsLint.kt [hunk 7]
```diff
@@ -51,6 +51,14 @@ enum class RsLint(
                 WARN -> ProblemHighlightType.WEAK_WARNING
                 else -> super.toHighlightingType(level)
             }
+    },
+
+    UnreachablePattern("unreachable_patterns", listOf("unused")) {
+        override fun toHighlightingType(level: RsLintLevel): ProblemHighlightType =
+            when (level) {
+                WARN -> ProblemHighlightType.WARNING
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/lints/RsLint.kt:59]: The default implementation does exactly the same things so you can drop `toHighlightingType` here at all

File: src/main/resources/META-INF/rust-core.xml [hunk 8]
```diff
@@ -521,10 +521,15 @@
                          enabledByDefault="true" level="ERROR"
                          implementationClass="org.rust.ide.inspections.RsTraitImplementationInspection"/>
 
-        <localInspection language="Rust" groupName="Rust"
-                         displayName="Match Check"
+        <localInspection language="Rust" groupPath="Rust" groupName="Lints"
+                         displayName="Useless match arm"
```
Comment (Undin) [inline src/main/resources/META-INF/rust-core.xml:525]:
```suggestion
                         displayName="Unreachable match patterns"
```

or even 
```suggestion
                         displayName="Unreachable patterns"
```

File: src/main/resources/inspectionDescriptions/RsUselessMatchArm.html [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<html>
+<body>
+Checks that <code>match</code> expression doesn't have useless (unreachable) patterns.
```
Comment (Undin) [inline src/main/resources/inspectionDescriptions/RsUselessMatchArm.html:3]:
```suggestion
Checks that <code>match</code> expression doesn't have unreachable patterns.
```

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            if (exprType.containsTyOfClass(TyUnknown::class.java)) return
+            try {
+                checkUselessArm(matchExpr, holder)
+            } catch (todo: NotImplementedError) {
+            } catch (e: CheckMatchException) {
+            }
+        }
+    }
+
+    private fun checkUselessArm(match: RsMatchExpr, holder: RsProblemsHolder) {
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt:41]: Maybe `checkUselessPatterns` is better?

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt]: I agree simple `WARING` highlighting type is better and more consistent with other languages:

image

My main concern was about using `registerProblem` instead of `registerLintProblem`

File: src/main/resources/META-INF/rust-core.xml [hunk 11]
```diff
@@ -521,10 +521,15 @@
...
+                         enabledByDefault="true" level="WARNING"
+                         implementationClass="org.rust.ide.inspections.checkMatch.RsUselessMatchArmInspection"/>
+
+        <localInspection language="Rust" groupPath="Rust" groupName="Lints"
```
Comment (Undin) [inline src/main/resources/META-INF/rust-core.xml:529]: Oh, I don't think we should move `RsNonExhaustiveMatchInspection` to `Lints` group. Only `RsUselessMatchArmInspection`. Yeah, they are very similar from implementation side but not from user point of view: non-exhaustive `match` is a compilation error and it even has own number in error index. At the same time, an unreachable pattern is warning by default and can be suppressed by user.
So let's revert this change

**Review by Undin — APPROVED** [review]
LGTM except minor comments

Thank you for this! It should make user experience more consistent

File: src/main/resources/META-INF/rust-core.xml (hunk 11 above)
Comment (ortem) [reply src/main/resources/META-INF/rust-core.xml:529]: Sure, I did it accidentally, thanks for your attention!

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/RsNonExhaustiveMatchInspection.kt (hunk 4 above)
Comment (ortem) [reply src/main/kotlin/org/rust/ide/inspections/checkMatch/RsNonExhaustiveMatchInspection.kt:34]: This name was given to match the original `check_exhaustive` method in rustc implementation

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt (hunk 6 above)
Comment (ortem) [reply src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt:24]: Let's use `RsUnreachablePatternsInspection`? It is located in `checkMatch` package so it's easy to discover that it is only about match expressions

Comment (Undin) [reply src/main/kotlin/org/rust/ide/inspections/checkMatch/RsUselessMatchArmInspection.kt:24]: 👍

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/RsNonExhaustiveMatchInspection.kt (hunk 4 above)
Comment (Undin) [reply src/main/kotlin/org/rust/ide/inspections/checkMatch/RsNonExhaustiveMatchInspection.kt:34]: ok

**Review by Undin — APPROVED** [review]
bors r+

**Comment by dima74** [comment]
Looks like this pull request also fixes #5901!
@ortem probably it is worth to add test from #5901
